Automattic/wp-calypso
# Email stats: skeleton for click stats

#### Proposed Changes

This PR lays the groundwork to bring click stats into email stats. It does a lot of refactoring to avoid any code duplication.

- `statsEmailOpensForPeriod` and `statsEmailOpensAllTime` in [wpcom.js's site.js](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-20b399f27d45d23194b00840a5c9333c9271b3400de0e028d28e9e79975eaf54) are now called `emailStatsForPeriod` and `emailStatsAllTime`. They now both take a `statType` argument, so you can differentiate between `opens` and `clicks`.
- Regarding [Redux actions](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-f2500d73aae7e42eb64c5349341c9fe9d87915a1088416e49897b87a9319c552):
  - `requestEmailOpensStats` is now called `requestEmailStats`, and also takes a statType.
  - `emailOpenStatsAlltimeTransform` and `emailOpenStatsPeriodTransform` do the same for both opens and clicks. Hence they are now called `emailStatsAlltimeTransform` and `emailStatsForPeriodTransform`.
  - `requestEmailPeriodStats` and `requestEmailAlltimeStats` are both simplified.
- No changes were needed in the Reducers 🥳
- In the [&lt;QueryEmailStats&gt; component](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-860cc7a5c61648f810c366d60238901cf7663c5d12bfedcb57ddbe2fb3ff3eef) statType is now also a property. 
- In [the site controller](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-71e9a204a637b410f5b0c4072fb0001b980183c95b390c1adfb92c1d8ea3df09), instead of duplicating the `emailOpen` function to accommodate clicks, I opted to create an `emailStats` function that takes the statType from the router
- &lt;StatsEmailOpenDetail&gt; is now called [&lt;StatsEmailDetail&gt;](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-6049e04828362101368d4c81a0da0a8329c83217eb30400ed042cfdd2cc6ad11) and contains some changes to accommodate multiple types of stats:
  - A separate function is introduced to get the charts: `getCharts` (in [client/state/stats/emails/utils.js](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-240061443075a8d2e94ffdebb027c8e81c225debe7544f1c3c2824c8705944ab). This cleans up the top of the file considerately.
  - `statType` is a prop on &lt;StatsEmailDetail&gt and is passed to all children requiring it.
  - The CSS class `stats__email-opens` is renamed to `stats__email-detail`.
- &lt;StatsEmailOpenTopRow&gt; is now called [&lt;StatsEmailTopRow&gt;](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-156a6c713e8b2d8d3bd8fe3939c088b839bd2193b858f667910b1458ea056e88) and now also requires `statType` to be passed.
- In [client/state/stats/lists/utils.js](https://github.com/Automattic/wp-calypso/compare/add/email-click-stats-skeleton?expand=1#diff-aac64af7e84d7057a1fc73e7036b3128b22fa6bdc55e8e82d5faca6505d5f06b) the results now get sorted, because the mock API doesn't return them sorted 🤷‍♂️ I also fixed some ESLint warnings in this file.

##### Breaking changes

The routes change with this PR:

/stats/email/open becomes /stats/email/opens

#### Testing Instructions

* Apply this PR
* Visit the Jetpack Stats and verify that clicking a post under "Email opens" still brings you to the opens stats: `http://calypso.localhost:3000/stats/day/<siteid>?flags=newsletter/stats`
* You should now be here: `http://calypso.localhost:3000/stats/email/opens/<siteid>/day/<postid>?flags=newsletter/stats` . This should still work!
* Replace `open` in the url to `click`: `http://calypso.localhost:3000/stats/email/clicks/<siteid>/day/<postid>?flags=newsletter/stats` . You should see (fake) click data.

<img width="1229" alt="CleanShot 2023-01-18 at 12 48 04@2x" src="https://user-images.githubusercontent.com/528287/213163871-df317d9a-20d7-4de0-8c57-0ec3b9b37760.png">

#### Pre-merge Checklist





Related to #
